File: src/test/java/com/koddy/server/file/utils/converter/FileConverterTest.java
package com.koddy.server.file.utils.converter;

import com.koddy.server.common.UnitTest;
import com.koddy.server.file.domain.model.RawFileData;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.web.multipart.MultipartFile;

import static com.koddy.server.common.utils.FileMockingUtils.createFile;
import static com.koddy.server.file.domain.model.FileExtension.PNG;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

@DisplayName("File -> FileConverter 테스트")
class FileConverterTest extends UnitTest {
    @Nested
    @DisplayName("파일 Convert")
    class ConvertFile {
        @Test
        @DisplayName("MultipartFile이 비어있으면 null이 반환된다")
        void returnNull() {
            // given
            final MultipartFile file = new MockMultipartFile("cat.png", new byte[0]);

            // when
            final RawFileData result = FileConverter.convertFile(file);

            // when - then
            assertThat(result).isNull();
        }

        @Test
        @DisplayName("MultipartFile -> RawFileData로 Converting한다")
        void success() {
            // given
            final MultipartFile file = createFile("cat.png", "image/png");

            // when
            final RawFileData result = FileConverter.convertFile(file);

            // then
            assertAll(
                    () -> assertThat(result.fileName()).isEqualTo("cat.png"),
                    () -> assertThat(result.contentType()).isEqualTo("image/png"),
                    () -> assertThat(result.extension()).isEqualTo(PNG)
            );
        }
    }
}

File: src/test/java/com/koddy/server/file/utils/converter/FileConverterTest.kt
package com.koddy.server.file.utils.converter

import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.utils.FileVirtualCreator.createFile
import com.koddy.server.file.domain.model.FileExtension
import com.koddy.server.file.domain.model.RawFileData
import com.koddy.server.file.exception.FileException
import com.koddy.server.file.exception.FileExceptionCode
import io.kotest.assertions.assertSoftly
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.throwable.shouldHaveMessage
import org.springframework.mock.web.MockMultipartFile
import org.springframework.web.multipart.MultipartFile

@UnitTestKt
@DisplayName("File -> FileConverter 테스트")
internal class FileConverterTest : DescribeSpec({
    describe("FileConverter's convertFile") {
        context("MultipartFile's content가 비어있으면") {
            val file: MultipartFile = MockMultipartFile("cat.png", ByteArray(0))

            it("FILE_NOT_UPLOADED 예외가 발생한다") {
                shouldThrow<FileException> {
                    FileConverter.convertFile(file)
                } shouldHaveMessage FileExceptionCode.FILE_NOT_UPLOADED.message
            }
        }

        context("MultipartFile's content가 존재하면") {
            val file: MultipartFile = createFile("cat.png", "image/png")

            it("RawFileData 도메인으로 변환한다") {
                val result: RawFileData = FileConverter.convertFile(file)

                assertSoftly(result) {
                    fileName shouldBe "cat.png"
                    contentType shouldBe "image/png"
                    extension shouldBe FileExtension.PNG
                }
            }
        }
    }
})
